app/src/main/java/org/oppia/android/app/home/HomeViewModel.kt

```diff
@@ -0,0 +1,185 @@
+import org.oppia.android.domain.profile.ProfileManagementController
+import org.oppia.android.domain.topic.TopicListController
+import org.oppia.android.util.data.DataProvider
+import org.oppia.android.util.data.DataProviders.Companion.combineWith
+import org.oppia.android.util.data.DataProviders.Companion.toLiveData
+import org.oppia.android.util.logging.ConsoleLogger
+import org.oppia.android.util.parser.StoryHtmlParserEntityType
+import org.oppia.android.util.parser.TopicHtmlParserEntityType
+import org.oppia.android.util.system.OppiaClock
+
+private const val PROFILE_AND_ONGOING_STORY_COMBINED_PROVIDER_ID = "profile+ongoingStoryList"
+private const val HOME_FRAGMENT_COMBINED_PROVIDER_ID = "profile+ongoingStoryList+topicListProvider"
+
+/** [ViewModel] for layouts in home fragment . */
+@FragmentScope
+class HomeViewModel(
+  private val activity: AppCompatActivity,
+  private val fragment: Fragment,
+  private val oppiaClock: OppiaClock,
+  private val logger: ConsoleLogger,
+  private val internalProfileId: Int,
+  private val intentFactoryShim: IntentFactoryShim,
+  private val profileManagementController: ProfileManagementController,
+  private val topicListController: TopicListController,
+  @TopicHtmlParserEntityType private val topicEntityType: String,
+  @StoryHtmlParserEntityType private val storyEntityType: String
+) : ObservableViewModel() {
+
+  private val profileId: ProfileId = ProfileId.newBuilder().setInternalId(internalProfileId).build()
+  private val promotedStoryListLimit =
+    activity.resources.getInteger(R.integer.promoted_story_list_limit)
+
+  private val profileDataProvider: DataProvider<Profile> by lazy {
+    profileManagementController.getProfile(profileId)
+  }
+
+  private val ongoingStoryListSummaryDataProvider: DataProvider<OngoingStoryList> by lazy {
+    topicListController.getOngoingStoryList(profileId)
+  }
+
+  private val topicListSummaryDataProvider: DataProvider<TopicList> by lazy {
+    topicListController.getTopicList()
+  }
+
+  private val homeItemViewModelListDataProvider: DataProvider<List<HomeItemViewModel>> by lazy {
+    // This will block until all data providers return initial results (which may be default
+    // instances). If any of the data providers are pending or failed, the combined result will also
+    // be pending or failed.
+    profileDataProvider.combineWith(
+      ongoingStoryListSummaryDataProvider,
+      PROFILE_AND_ONGOING_STORY_COMBINED_PROVIDER_ID
+    ) { profile, ongoingStoryList ->
+      listOfNotNull(
+        computeWelcomeViewModel(profile),
+        computePromotedStoryListViewModel(ongoingStoryList)
+      )
+    }.combineWith(
+      topicListSummaryDataProvider,
+      HOME_FRAGMENT_COMBINED_PROVIDER_ID
+    ) { homeItemViewModelList, topicList ->
+      computeAllTopicsItemsViewModelList(topicList).let {
+        if (it != null) {
+          homeItemViewModelList + it
+        } else {
+          homeItemViewModelList
+        }
+      }
+    }
```

Comment:
I think this is the right idea, but might it be simpler if we had the computeAllTopicsItemsViewModelList() function return empty in the case it would otherwise be null? I think that simplifies this whole block to just a list concatenation.

Reply:
Done -- I don't particularly like returning an empty list since it seems like an inaccurate representation of what's happening but I do agree that the simplification is more useful.